eduroam database v2 adaptation

zmousm opened the pull request:
Implement #62 

~~WIP. Models and migrations are mostly done. Still pending forms, views, templates management commands.~~ Now complete.

Important notes, caveats:
* `UUID` is enforced for `Institution.instid`, `ServiceLoc.locationid`. Such fields are not editable through forms, while `parse_institution_xml` defaults to deriving a `UUID` by means of MD5 hashing, if the input is unsuitable.
* `Coordinates` are modeled as m2m for `ServiceLoc`, but only a single set is supported currently. This is enforced by a signal receiver. Longitude, latitude `ServiceLoc` cached properties work as a proxy, together with a related mechanism in `ServiceLocForm`. On the other hand, `parse_institution_xml` retains the first set of coordinates it encounters in a location.
  * Therefore altitude is currently ignored.
* `Coordinates` are currently ignored in institution edit form, nor are they otherwise presented anywhere, while they are processed by `parse_institution_xml` as expected.
* Data migration from `ServiceLoc.wired` to `ServiceLoc.wired_no` is driven by `settings.SERVICELOC_DERIVE_WIRED_NO`, which defaults to mapping `True` to 42.
* Compatibility with all eduroam database versions is retained throughout db views and `parse_institution_xml`. For the latter the version may be provided through a command-line option. For the former the requested version is inferred by considering path, query parameter or resource (realm vs. ro). If no version is specified/hinted, a default is used, which, as of now, is version 2.
  * This should probably be documented in release/upgrade notes.

@vladimir-mencl-eresearch,@c00kiemon5ter, @ioparaskev please start reviewing

zmousm committed:
Add choices for EDB2, using some lightweight enum-like objects

* Add function to produce lightweight enum-like objects using
  collections.namedtuple (substitute for enum types in PY3 and Django 3)
  * Attribute access only returns the value, so it can be used for
    choice defaults
  * Only use this (currently) for choices where a default is called out
* Map the EDB2 string value in ERTYPES with a function to translate from
  the integer value, which is stored in the DjNRO database

zmousm committed:
Add Address_i18n model

zmousm committed:
Add InstitutionDetails fields for EDB2

* venue_info: use simple, custom validate_venue_info() validator
  * django.core.validators.validate_comma_separated_integer_list() seems
    redundant
* Keep number_user, number_id fields for now, for EDBv1 backwards compat

zmousm committed:
Add Institution fields for EDB2

* Add inst_name, also keeping org_name
  * Reverse generic relation does not affect the db schema, so let's
    keep both for backwards compat

zmousm committed:
Add Contact fields for EDB2

zmousm committed:
Add Realm fields for EDB2

* ROid default taken from settings.NRO_ROID (no fallback) and it is
  required to be unique
  * Assume existing databases will have a single Realm object
* stype hardcoded to 0 (FLRS), removed from database but kept as an
  attribute for now

zmousm committed:
Add ServiceLoc fields for EDB2

zmousm committed:
Clean up and realign MultiSelectField, MultiSelectFormField

* Fix get_prep_value() and to_python() to return proper string, list
  values respectively
* Remove get_db_prep_value(), as overriding it is unnecessary
* Have contribute_to_class() provide a function adapted from
  django.db.models.Model._get_FIELD_display() and mimic the parent for
  attaching it to the model using curry()
* Have validate() call the overridden method for each value
* Have formfield() only override the form class with our own
* Let MultiSelectFormField only override the form widget and remove
  extra args by means of introspection
  * extra args for TypedChoiceField added by calling parent
    models.Field.formfield()
* Remove redundant overridden method value_to_string()

zmousm committed:
Let the database value separator for MultiSelectField be configurable

zmousm committed:
EDB2 schema migration, part 1

zmousm committed:
Fix previous migration

* Let Address_i18n defer lang choices to settings
* Let Realm defer roid default to settings
* Properly backfill UUID values

zmousm committed:
Add Coordinates model and relationships

* Add Coordinates model with latitude, longitude and optional altitude
* Add nullable FK to Coordinates from Realm and InstitutionDetails
* Add django-sortedm2m to requirements.txt, as the provider for our
  sorted ManyToMany fields
* Add sortedm2m to INSTALLED_APPS, as necessary for accessing the
  provided form widget
* Add coordinates as a SortedManyToMany field in ServiceLoc
  * Ensure only a single coordinates object can be added, for the time
    being, using a m2m_changed signal receiver
    * This ensures enforcement even in management commands, yet form
      validation should also be added to raise meaningful errors

zmousm committed:
EDB2 schema migration, part 3: Cleanup

zmousm committed:
EDB2 schema migration, part 2: Coordinates

zmousm committed:
EDB2 data migration, part1

* Use AppAwareRunPython rather than plain RunPython
* Create Address_i18n(lang='en') objects from address_street,
  address_city, iterating on objects of the respective models
  * Reverse migration sets the latter fields from the first related
    Address_i18n object, by using a presumably much faster, single
    statement embedding a Subquery expression (Django 1.11 new feature)
    with LIMIT 1 trick -- this will return NULL if no linked objects,
    which should raise an error
* Populate ServiceLoc.tag from the respective legacy boolean fields
  * Reverse migration reflects the existence of a tag in the respective
    boolean field value
  * This is slow, as each object is individually saved, but it can work
    more efficiently with bulk_update() in Django >= 2.2
* Beware that
  * ContentType *must* be retrieved from the app registry, or weird
    mismatch errors will creep in:
    Cannot query "Realm": Must be "ContentType" instance
  * Reverse GenericRelation manager apparently is not registered on
    instances of models retrieved from the registry

zmousm committed:
EDB2 data migration, part 2: Coordinates

* Use AppAwareRunPython rather than plain RunPython
* Create Coordinates objects from latitude, longitude, iterating on
  ServiceLoc objects
  * update_or_create() matches only on the ServiceLoc relation,
    unconditionally updating lat, lon
    * Multiple objects matched should raise an error
  * The resulting Coordinates object replaces the whole set of objects linked to
    the respective ServiceLoc object
* Reverse migration sets the latter ServiceLoc fields from the first
  linked Coordinates object, by using a single statement embedding a
  Subquery expression with LIMIT 1 trick -- this will return NULL if no
  linked objects, which should raise an error

zmousm committed:
Introduce edumanage.migrations.AppAwareRunPython

This wraps RunPython so as to auto-provide the app_label as an extra
argument to migration code.

zmousm committed:
Use AppAwareRunPython in migrations/0005_edb2_schema_part1.py

zmousm committed:
EDB2 schema cleanup: Remove redundant fields

* address_street, address_city
* ServiceLoc tag fields
  * Also remove from edumanage.admin, replace with ServiceLoc.tag
* latitude, longitude

zmousm committed:
Reinstate ServiceLoc.{latitude,longitude} as cached property
descriptors

* Introduce utils.functional.cached_property, like
  django.utils.functional.cached_property with the addition of setter
  and deleter
* Re-add ServiceLoc.{latitude,longitude} as cached properties that
  * fetch the respective value from the first linked Coordinates
    object and cache it in the ServiceLoc instance dict
  * save an assigned/updated value in the cache
  * pop the value from the cache upon deletion and also remove the
    first Coordinates object matching the respective value
* Add a ServiceLoc post_save signal receiver that creates or updates a
  linked Coordinates object with cached values, if found and not null,
  and pops them from the cache, so as to ensure that subsequent
  property access will fetch the persisted value

zmousm committed:
Fix Address_i18n.__str__ and add get_address_factory()

* Add Address_i18n.get_address_factory() like
  Name_i18n.get_name_factory()
  * Let it fallback to a default language (by default)
    * if no lang is specified
    * if nothing is returned for the specified lang
* Have Address_i18n.__str__() build an address by joining non null/empty
  street, city fields
* Add a property knob for including language in the stringified address
* Have get_address_factory() turn off such knob
* It should be noted that no other means for overriding __str__ on an
  Address_i18n instance is effective (or elegant)
* An Address_i18n instance returned by get_address_factory() can thus
  be used both for access to its' fields and as a string

vladimir-mencl-eresearch commented:
Hi @zmousm , thanks a lot for this work!

I'm just about to go on Christmas leave (for southern hemisphere/New Zealand, these are the main summer holidays) and I'll be back on deck from January 13th.

I'll see if I manage to review anything in the last ~ hour I have in the office, but I'm afraid in won't be much - I promise to look at this once I'm back in the week of January 13th.

Cheers,
Vlad

zmousm commented:
No worries @vladimir-mencl-eresearch, happy holidays!

ioparaskev commented on edumanage/models.py:
```diff
@@ -21,109 +42,157 @@ class MultiSelectFormField(forms.MultipleChoiceField):
+        for key in list(kwargs):
+            if key not in supercls_init_args:
+                del kwargs[key]
+        supercls.__init__(*args, **kwargs)
```
maybe instead of iterating and deleting keys from the dict, you could achieve the same thing with:
```python
supercls.__init__(*args, **{key,val for (key,val) in kwargs.items() if key in supercls_init_args})
```

zmousm replied:
Sure. I thought it was better readable like that.

ioparaskev replied:
main reason I proposed this was to avoid having `del`

ioparaskev commented on edumanage/models.py:
```diff
@@ -21,109 +42,157 @@ class MultiSelectFormField(forms.MultipleChoiceField):
+    def deconstruct(self):
+        name, path, args, kwargs = super(
+            MultiSelectField, self).deconstruct()
+        if self.separator != self._separator_default:
```
since we always want to have the `kwargs['seperator'] == self.seperator` why not make the assignment in the first place? I'm not sure there's a point having the `if` statement here unless I'm missing something

zmousm replied:
Django docs [suggest the deconstructor should only return kwargs that don't carry the default value](https://docs.djangoproject.com/en/1.11/howto/custom-model-fields/#field-deconstruction) (as they are deemed redundant otherwise).

zmousm committed:
Install and enable widget_tweaks django app/plugin

The set_data template tag it provides will be used in edumanage
templates.

zmousm committed:
Simplify edumanage.templatetags.tolocale with simple_tag()

And add variant tolocale_get with assignment_tag()

zmousm committed:
Let tolocale accept an optional parameter to determine the method to
call

zmousm committed:
Clean up edumanage templates and respective views

* service_details.html does not render form, thus remove all refs to
  that object
* institution.html does not render form, thus remove form from
  edumanage.views.institutions()

zmousm committed:
Use django.utils.inspect.getargspec for MultiSelectFormField kwargs
inspection

And accommodate PY2/PY3 differences in returned values

zmousm committed:
Align utils.functional.cached_property with Python property equiv

* Store respective functions under fget, fset, fdel properties
  * Django QuerySet checks this to whitelist property descriptors for
    lookup
* Check fget before setting self.__module__
* Raise AttributeError if no getter is given
* Let set and delete return None
* Let delete inspect fdel args so as to call it with or without a value
  argument

zmousm committed:
Deduplicate edumanage.forms

* NameFormSetFact becomes Generici18nFormSetFact
  * it can be used both in Name_i18n and Address_i18n formset factories
  * a short description of the object as used is given as a kwarg to the
    constructor
* Remove unused code in UrlFormSetFact.clean
* Clean up UrlFormSetFactInst

zmousm committed:
edumanage.views.add_{services,institution_details}: Deduplicate and
adapt for Address_i18n

* Deduplicate formset generation to blocks driven by formset_params for
  any number of formsets
  * Use Generici18nFormSetFact for both name and url forms
    * Use partialclass() to pre-supply the respective obj_descr value
  * Chain generic_inline_formset_factory() with formset creation
* Deduplicate main form and auxiliary field generation for different
  code paths
  * with or without existing {InstitutionDetails,ServiceLoc} object
  * GET or POST

zmousm committed:
Adapt templates for Address_i18n: Get address in LANGUAGE_CODE with
tolocale_get

* edumanage/institution.html
* edumanage/services.html
* overview/index.html
* front/participants.html

zmousm committed:
service_details.html: Adapt for Address_i18n

* Replace fixed address_street, address_city fields with loop over
  address objects

zmousm committed:
institution_edit.html: Adapt for Address_i18n

* Replace fixed address_street, address_city fields with tabular form
  for any number of address objects

zmousm committed:
services_edit.html: Adapt for ServiceLoc.tag

* Replace fixed tag fields with MultiSelectField form widget
  (CheckboxSelectMultiple)

zmousm committed:
service_details.html: Adapt for ServiceLoc.tag

* Replace fixed tag fields with ServiceLoc().get_tag_display()

zmousm committed:
services_edit.html: Adapt for Address_i18n

* Replace fixed address_street, address_city fields with tabular form
  for a number of address objects
* Use widget_tweaks set_data template tag to tag street, city, lang
  fields with identifying data attributes
* Adapt javascript to do reverse geocoding queries for each row in the
  respective language and set the values of the data-tagged form
  controls
  * Group reverse geocoding queries in rows per language, for efficiency
  * However Google Maps Javascript API only supports setting language
    upon load, not per request. So only do reverse geocoding for
    English, matching the hardcoded language parameter to the API script
    URL, but keep the code in place for future reference.
  * Also do not clear the value of address fields before updating them.

zmousm committed:
edumanage.views.instxml: Adapt for ServiceLoc.tag

* Either produce the EDB2 tag element or
* Recreate boolean EDB1 tag elements for tags included in ServiceLoc tag
  values

zmousm committed:
edumanage.views.{ourPoints,localizePointNames}: Adapt for Address_i18n

* Replace fixed address_street, address_city fields with loop over
  address objects; group by and flatten by language
* Similar treatment is probably necessary for all i18n objects

zmousm committed:
edumanage.views.ourPoints: Adapt for ServiceLoc.tag

* Recreate fixed tag dict members with loop over
  ServiceLoc.LOCATION_TAGS comparing to ServiceLoc tag values

zmousm committed:
edumanage.views.ourPoints: Replace ServiceLoc.wired with wired_no

Also update affected js

zmousm committed:
edumanage.views:{inst,realm}xml: Deduplicate and adapt for Address_i18n

* Add simple eduroam db versioning (temporarily hardcoded)
* Introduce xml_address_elements()
  * Inject address elements into the output xml, as required for the
    respective version schema
  * Deduplicate all such code from the views

zmousm committed:
edumanage.management.commands.parse_institution_xml: Adapt for
ServiceLoc.tag

Differentiate between parsing
* boolean tags for EDB1 or
* the tag element for EDB2

zmousm committed:
edumanage.management.commands.parse_institution_xml: Replace
ServiceLoc.wired with wired_no

zmousm committed:
edumanage.management.commands.parse_institution_xml: Adapt for
Address_i18n

* Add simple eduroam db versioning
* Introduce parse_and_create_address()
  * It handles harvesting address elements to Address_i18n objects
* Refactor parse_and_create_{institution,serviceloc}()
  * Delegate all address processing to the former method
  * Ensure sub-elements always carry a lang attribute (eduroam db v2
    compliant) to simplify parsing elements by the former method

zmousm committed:
Introduce utils.functional.partialclass

And backfill partialmethod() for PY<3.4

zmousm committed:
fixup! Let the database value separator for MultiSelectField be configurable

Fix typo: self.separator -> field.separator

zmousm committed:
MultiSelectField: override get_choices() to disable include_blank if
blank is allowed by the field instance

* blank_choice does not make sense for our CheckboxSelectMultiple widget
* Adjust edumanage.views.ourPoints(), removing test for blank choice

zmousm committed:
Let MultiSelectField.from_db_value() handle both blank and None as
special cases

Avoid returning an empty database value as ['']

zmousm committed:
edumanage.forms.ServiceLocForm: Derive from ModelFormWithPropertyFields
and declare proxy fields for latitude, longitude

* Use django.forms.models.fields_for_model() to declare latitude,
  longitude fields in ServiceLocForm based on the corresponding
  Coordinates model fields
* Manually list such fields in _property_fields, so that the base class
  (ModelFormWithPropertyFields) constructor and save methods can fetch
  and persist the corresponding values to/from the form through the
  respective ServiceLoc cached properties
  * This manual list is a necessity in order to avoid excessive
    complexity; otherwise one would need to override both
    ModelFormMetaClass.__new__ and BaseModelForm.{__init__,save} in
    order to do this properly
* Exclude coordinates (otherwise required) from ServiceLocForm fields

zmousm committed:
edumanage.management.commands.parse_institution_xml: Streamline parsing
multi-value elements

zmousm committed:
edumanage.views.instxml: Produce string institution type for EDB2

zmousm committed:
edumanage.management.commands.parse_institution_xml: Parse string
institution type for EDB2

* Add edumanage.models.get_ertype_number(), providing the reverse lookup
  of get_ertype_string()

zmousm committed:
edumanage.management.commands.parse_institution_xml: Parse coordinates

* Don't create a parse_and_create method as parsing, input validation,
  object creation are dispersed and somewhat different for institution
  vs. location
* Enumerate Coordinates fields to parse the respective elements
* Pack values into dict(list) or list(dict) seqs respectively for EDB1
  and EDB2
* Skip a location element if multiple coordinates are encountered
  (unsupported currently)
* Chain filter ServiceLoc objects with Coordinates parameters
* Reorganize code handling EDB1/EDB2 required tags

zmousm committed:
edumanage.views.{inst,realm}xml: Render coordinates

* Introduce xml_coordinates_elements()
  * Find related Coordinates objects, handling both FK (and None) as
    well as M2M from the given obj
  * Enumerate Coordinates fields to render the respective elements
  * Deduplicate all such code from the views

zmousm committed:
edumanage.views.instxml: Differentiate institution name tag based on EDB
schema version

zmousm committed:
MultiSelectField: Deduplicate and sort values to/from the database

zmousm committed:
edumanage.management.commands.parse_institution_xml: Refactor, parse
remaining EDB2 data

* Parse EDB2 tags
* Parse instid, locationid, deriving UUID if an unsuitable value is
  provided
  * MD5 hash provides a suitable 32 digit hex value
* Map tags to respective model fields through the convention we use so
  far: model fields specify a db_column value matching the respective
  EDB schema field that is different from the field name
* ServiceLoc coordinates: keep the first set rather than skip the whole
  location element, if more are provided (unsupported currently)
* Look for an existing ServiceLoc object based only on institutiond and
  locationid (and coordinates)
  * fields are updated for other parameters after an object is matched
* Let the name comparison extension sort the datasets and also consider
  lang, and only prepare it if the previous query matches
* Consider instid to match an existing Institution object, pre-empting
  the check for matching Institution names
* Simplify updating an existing InstitutionDetails object

zmousm committed:
edumanage.management.commands.parse_institution_xml: Parse the proper
institution name tag for EDB2

zmousm committed:
edumanage.management.commands.parse_institution_xml: Let deriving UUIDs
be optional

zmousm committed:
Add ERTYPE_ROLES, ERTYPES membership testing

* Categorize ERTYPES values by role into ERTYPE_ROLES, so such
  categories can be called out by name
* Override Choices.__contains__ for namedtuples, so one can test
  e.g. for value in ERTYPES

zmousm committed:
edumanage.views.instxml: Render remaining EDB2 data

zmousm committed:
edumanage.views.realmxml: Add remaining EDB2 fields

* Also differentiate root elements based on EDB schema version

zmousm committed:
edumanage.views.instxml: Speed up with joins and groupby

* Prefetch all related Institution, InstitutionDetails objects, except
  for ServiceLoc
* Do the same for ServiceLoc, independently
* Order and group ServiceLoc objects by institutionid
* Look up the Institution PK in the resulting dict and iterate the list
  of ServiceLoc objects
  * Store such a list in the dict rather than the (shared-source) group
    iterator

zmousm committed:
edumanage.views.instxml: institution.institutiondetails is not None if
unset

Remove redundant check

zmousm committed:
edumanage.views.instxml: Render UUIDs as hex strings

Without dashes

zmousm committed:
edumanage.views.instxml: Do not render optional elements if blank/None

zmousm committed:
Introduce DelimitedValueExactLookup for MultiSelectField

* Look up the delimiter in field.separator
* Expand an exact lookup to alternatively also perform a contains,
  prefix and suffix match so that a given rhs may match an individual
  value, stored anywhere between, before or after a delimiter,
  respectively, in the charfield db value
* Temporarily override self.lookup_name so as to fill in the sql
  operator by using built-in facilities

zmousm committed:
Reference ERTYPES, ERTYPE_ROLES symbolically rather than by value

Both for equality and membership tests

zmousm committed:
extras/servdata_consumer.py: Emulate symbolic ref for ERTYPE_ROLES

Through a const class, so as to avoid importing from edumanage.models

zmousm committed:
Add eduroam db versioning

* Install and use semantic_version.Version
  * Prefer this package as it implements coercing from arbitrary
    (non-semantic) version strings
* Add utils.edb_versioning
  * EduroamDatabaseVersionDef base class holds EDB version definitions
  * EduroamDatabaseVersion is a sub-class of the former
    * It is instantiated through syntaxes most commonly supported by
      semantic_version.Version, including coercing
    * It validates that the derived version is among the explicitly
      allowed versions
    * It provides short-hand properties to compare with the default
      and all configured versions
    * It also acts as a proxy to semantic_version.Version by means of
      metaclass DelegateVersion, so that methods and props defined in
      the former (such as for equality/comparison tests) can be used
      directly on its' instances (rather than through self.version)
      * This is impossible to do with inheritance, as returning Version
        instances is hardcoded in various class methods etc. therein
  * LazyEDBVersion is a django.utils.functional.SimpleLazyObject wrapper
    of the former, which defers settings lookup for the default version
  * DEFAULT_EDUROAM_DATABASE_VERSION holds a lazily instantiated
    EduroamDatabaseVersion object, by instructing LazyEDBVersion where
    to look for the default version in django.conf.settings (also lazy)
    * Such mechanics are necessary so as to avoid import loop between
      settings and this module
  * edb_version_from_request() is a helper that attempts to derive the
    EduroamDatabaseVersion from a match on the request path, a request
    parameter or a match on the request resource, raises an exception
    upon conflicting hints, and falls back to
    DEFAULT_EDUROAM_DATABASE_VERSION otherwise
    * This mimics DRF versioning classes to some extent
* Add edumanage.decorators.detect_eduroam_db_version(), which decorates
  EDB export views so as to call them with a EduroamDatabaseVersion
  parameter, derived through the previous helper function, or raise
  HTTP 400 if that fails
* Annotate EDB export urls with path and resource matching for use by
  the previous helper function
* edumanage.views.{inst,realm}xml: decorate with
  detect_eduroam_db_version() and replaces all comparisons to arbitrary
  (temp) EDB integer values with symbolic refs to EduroamDatabaseVersion
  instance properties: is_version_1, is_version_2, ge_version_2
* edumanage.management.commands.parse_institution_xml: Also replace
  arbitrary (temp) EDB integer values with respective symbolic refs, and
  have edb_version argument hold a EduroamDatabaseVersion instance and
  default to DEFAULT_EDUROAM_DATABASE_VERSION
* Add djnro.settings.EDUROAM_DATABASE_VERSIONS, a dict pointing to
  EduroamDatabaseVersionDef attributes for the default and allowed EDB
  versions
  * As of now, and unless overridden in local_settings, we allow EDB
    versions 1 and 2 and default to version 2

zmousm committed:
edumanage.{forms,views}: Refactor formsets

* Replace Generici18nFormSetFact with (even more generic) i18nFormSet
  * Besides obj_descr it also takes a required_value_sets attribute or
    init kwarg
  * clean() does not check for an empty formset, this must be handled
    with built-in facilities: min_num and validate_min
  * required_value_sets is a sequence of field-value combos that must be
    found in forms' data
    * Each field and value must be accompanied by a description, which
      can be marked for (lazy) translation, in such a structure: (field,
      field_description), (value, value_description)
  * A ValidationError is raised for every required value set not found
    in the formset
* Add i18nFormSetDefaultLang, a sub-class of the former which requires
  that a lang attribute with a value matching the default language
  (English) is included in the formset
* Add URL_i18nFormSet, also a sub-class which combines the lang
  attribute requirement with iteration over URL_i18n.URLTYPES values for
  urltype
* Use the former for formsets in add_institution_details, add_services
  views, along with the appropriate min_num:
  * 2 for institution URLs (info and policy, one of each)
  * 1 for addresses
    * This is optional for locations in EDB2, but we opt to require it
      so as to ensure backwards compatibility

zmousm committed:
fixup! Add ServiceLoc fields for EDB2

* Provide help_text for venue_info, physical_avail, operation_hours
* Add blank=True to AP_no, wired_no, as the formfield widget is marked
  as required otherwise

zmousm committed:
fixup! EDB2 schema migration, part 1

* Include help_text in
  * institutiondetails.venue_info
  * serviceloc
    * operation_hours
    * physical_avail
    * venue_info
* Add WPA3/AES to serviceloc.enc_level choices
* Add blank=True to serviceloc
  * AP_no
  * wired_no

vladimir-mencl-eresearch commented:
Hi @zmousm ,

Great job.  I've just gone through all the commits.  I'll comment inline, but it's mostly requests for clarification / more detailed documentation - no real issues discovered, looks good to me.

Cheers,
Vlad

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -100,30 +101,71 @@ def value_to_string(self, obj):
-    (2, 'SP only'),
-    (3, 'IdP and SP'),
+# https://www.djangosnippets.org/snippets/2402/
+def get_namedtuple_choices(*choices_tuples):
```
Hi @zmousm ,

I think it would be useful to document the required structure of the choices_tuples passed to this call.

From the code here, I can each tuple in the choices should start with an identifier, but I see the structure then quite varies:
* ER_TYPES has a numeric ID, others don't.
* ER_TYPES has has a short handle *after* a descriptive name, others reverse this order.

I think it would be good to define the structure to be passed - and then adhere to it.

zmousm replied:
Hello @vladimir-mencl-eresearch,

the structure is the sequence of iterables that [Django accepts for choices](https://docs.djangoproject.com/en/1.11/ref/models/fields/#choices), and in each of them the two items are preceded by the [field name for namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple). _ERTYPES is special in that each iterable also includes at the end the string value, which is new in eduroam db v2, so as to help with the mapping. However any such further items are ignored by `get_namedtuple_choices`.

Coming up, I will be adding some documentation.

vladimir-mencl-eresearch replied:
Agree that when I treat `ER_TYPES` as a separate special case, the structures look more uniform - a tuple of 3-tuples, each starting with an identifier.

Still, sometimes, the second field in each tuple (the internal value to store?) is a string, sometimes a number.  I assume the last field in each tuple is a description.

Having this documented would help a lot - thanks a lot in advance for that.

zmousm replied:
> Still, sometimes, the second field in each tuple (the internal value to store?) is a string, sometimes a number. I assume the last field in each tuple is a description.

That is correct, per the Django spec.

vladimir-mencl-eresearch replied:
Thanks for the patience with me - happy with just pointing to Django doc.

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -100,30 +101,71 @@ def value_to_string(self, obj):
+    ('IDPSP', 3, 'IdP and SP', 'IdP+SP'),
+)
+ERTYPES = get_namedtuple_choices(*_ERTYPES)
+def get_ertype_string(ertype):
```
This function should document the input parameter.  I assume it is the numeric ID, but it should be documented.

zmousm replied:
Alright

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -222,6 +320,27 @@ def __str__(self):
 
 
+@python_2_unicode_compatible
+class Address_i18n(models.Model):
```
I understand the structure is driven by what eduroam.org asked for ... but shouldn't an address also specify a Country?
Street and City does not look sufficient to me in the global context....

zmousm replied:
One might argue this can be derived considering the RO to which the particular object is linked and the country (area) they serve. I do get your point though, yet I am focused on covering eduroam db v2 schema changes in this (quite large) change set, and this can be added later.

vladimir-mencl-eresearch replied:
Sure, no problem.  I understand this is driven by the V2 structure and we can't change it easily from here - still wanted to raise it as an usability issue.  But that's not to be solved here, so I'm marking this part as resolved.

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -520,6 +716,12 @@ class InstitutionDetails(models.Model):
+        blank=True,
+        validators=[validate_venue_info],
+        db_column='inst_type'
+    )
```
Why is the field called`venue_info` but stored in db_column `inst_type` ?  That just looks like unnecessary complexity - can't we just use one name for both?

I expect there is a reasonable answer to this - just asking because it struck me as odd.

zmousm replied:
The eduroam db spec, both the abstract version and the XML (JSON) schema, defines some quite generic field names, which felt too confusing (or even technically inappropriate) to use for our Django model fields. We have always followed the convention of using the actual name in the spec for `db_column`, if at all possible, as this is hidden from casual usage, but we still get to record the mapping in there. And now we will actually get to use this mapping, as you will see in upcoming commits. I do understand however that as new eduroam db versions arise, such a mapping mechanism is inadequate and should be replaced by something that also considers the EDB version. I could not tackle this at this stage, though.

vladimir-mencl-eresearch replied:
Sure - also happy with this - glad to hear you did consider this carefully.

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -427,30 +546,64 @@ class ServiceLoc(models.Model):
-    address_city = models.CharField(max_length=64)
+    # TODO: multiple addresses can be specified [...] address in English is required
+    address = fields.GenericRelation(Address_i18n)
+    venue_info = models.CharField(
```
Same comment as for InstitutionDetails.venue_info:

> Why is the field called`venue_info` but stored in db_column `inst_type` ?  That just looks like unnecessary complexity - can't we just use one name for both?
> 
> I expect there is a reasonable answer to this - just asking because it struck me as odd.

zmousm replied:
See previous

vladimir-mencl-eresearch replied:
Agreed.

vladimir-mencl-eresearch commented on edumanage/models.py:
```diff
@@ -184,6 +264,24 @@ class Meta:
 
 
+@python_2_unicode_compatible
+class Coordinates(models.Model):
```
I think it would be helpful to document the units usedhere.

I assume it's degrees as float for lat/long - and meters for alt ... but I think it should better be documented.

zmousm replied:
Right. I will try to add some `help_text` in some fields.

vladimir-mencl-eresearch replied:
Thanks, looks all good now.

vladimir-mencl-eresearch reviewed (changes requested):
And that's it.

Note that I did not review: the changes to `MultiSelectField` / `MultiSelectFormField` (a bit too much outsideof  my area) - but it appears to have been reviewed extensively by @ioparaskev.

So once more, congratulations on the job well done.

Cheers,
Vlad

zmousm committed:
jquery.formset.js: Enable delete on all forms

Undo (comment-out) our modifications, reverting to upstream

zmousm committed:
edumanage.forms.InstDetailsForm: Mark error messages for translation

zmousm committed:
Add edumanage.forms.URL_i18nForm and use it to restrict ServiceLoc
urltypes

* URL_i18nForm restricts accepted urltype choices per valid_urltypes
  attribute or init kwarg
* edumanage.views.add_services: Use the former form class for the URL
  formset, rather than overwriting the urltype field in URL forms

zmousm committed:
Expose remaining EDB2 data in templates

* edumanage/add_user.html: Render type, privacy form fields
* edumanage/institution.html: Render instid, venue_info fields
* edumanage/institution_edit.html
  * Render venue_info form field
  * Render a read-only field for instid
    * Not an InstitutionDetails attribute, so not part of the form
      anyway
  * Separate non-form errors with line break, rather than joining them
    with comma
* edumanage/service_details.html: Render locationid, stage, geo_type,
  physical_avail, operation_hours, venue_info, wired_no fields
  * Hide geo_type for now
  * Replace wired
  * Move SSID further down
* edumanage/services_edit.html
  * Render stage, geo_type, physical_avail, operation_hours, venue_info,
    wired_no form fields
    * Hide geo_type for now (while coordinates remains a singleton)
    * Replace wired
  * Move SSID further to the end of the form
  * Separate non-form errors with line break, rather than joining them
    with comma
  * Render a read-only field for locationid
    * Render the value if the form is bound to a ServiceLoc instance
      (edit=True)
* Style read-only input fields with opacity resembling browser styles
  for disabled fields

zmousm committed:
edumanage.views.add_services: Do not require contact field

zmousm committed:
edumanage.models: Document get_namedtuple_choices, Choices methods,
_ERTYPES and get_ertype_string

zmousm committed:
fixup! Add Realm fields for EDB2

* Add NRO_ROID in local_settings.py.dist

zmousm committed:
fixup! EDB2 data migration, part 1

* Add additional data migration function for ServiceLoc wired <-> wired_no
  * Instrument it via settings.SERVICELOC_DERIVE_WIRED_NO
* Use attr lookup rather than getattr for settings

zmousm committed:
fixup! EDB2 schema cleanup: Remove redundant fields

* Also remove ServiceLoc.wired

zmousm committed:
fixup! EDB2 schema migration, part 3: Cleanup

* Include removal of serviceloc.wired

zmousm committed:
fixup! Reinstate ServiceLoc.{latitude,longitude} as cached property descriptors

* Handle Coordinates.MultipleObjectsReturned (which should never
  happen with db access through the ORM)

zmousm committed:
fixup! Add InstitutionDetails fields for EDB2

* Provide help_text for venue_info

zmousm committed:
fixup! Add choices for EDB2, using some lightweight enum-like objects

* Remove duplicate LOCATION_TYPES, covered by ServiceLoc.GEO_TYPES
* Replace PHYSICAL_AVAILABILITY_STATES with ServiceLoc.PHYSICAL_AVAILABILITY_STATES
* Add WPA3/AES to ServiceLoc.ENCTYPES

zmousm committed:
fixup! Add Coordinates model and relationships

Provide help_text for longitude, latitude, altitude

zmousm committed:
fixup! EDB2 schema migration, part 2: Coordinates

Include help_text in coordinates
  * longitude
  * latitude
  * altitude

zmousm commented:
I am now pushing the second round of commits (hopefully there won't be a third round).

The first 12 commits (prefixed with _fixup!_) are meant to be squashed into their preceding counterparts before merging. I am not doing this now so as to not confuse you and not force you to review everything all over again.

@vladimir-mencl-eresearch, @c00kiemon5ter, @ioparaskev please review.

I hope to find some time to update the high-level overview of the changes in the linked issue.

zmousm committed:
Add utils.edb_versioning.edb_version_fromto_resource and associated
template tag

* Map an eduroam database realm URL resource ("realm" or "ro") to an
  EduroamDatabaseVersion or vice versa.
* The template tag performs the latter, for the
  DEFAULT_EDUROAM_DATABASE_VERSION if no version is given, so as to
  provide the resource component for reversing the url for realmxml

zmousm committed:
services_edit.html: Fix longitude, latitude fields

* Remove (long forgotten) misplaced form.longitude.help_text
* Regroup longitude, latitude input fields with their errors and
  help_text widgets
  * This pushes each field/widget to their own line...
* Enable updatemap on keyup/change events from the former fields

zmousm committed:
Add and use template filter join_with_linebreaks

Replace joining errors with line breaks rather than comma, where
appropriate

zmousm committed:
Replace tolocale_get assignment template tag (deprecated) with
tolocale simple tag

zmousm commented:
I pushed 4 more commits to fix errata

zmousm commented:
I should also note that I have not effectively tested `parse_institution_xml` as of yet...

zmousm committed:
edumanage.views: Fix groupby expressions

* Skip redundant import
* Name group items in iteration

zmousm committed:
djnro.settings._dictmerge: Fix for PY3

zmousm committed:
fixup! fixup! EDB2 schema migration, part 1

* Do not use byte strings (inadvertently)
* Update venue_info help_text

zmousm committed:
fixup! edumanage.management.commands.parse_institution_xml: Parse coordinates

* Fix typo
* Only chain filter by coordinates for < EDB2
* Only compare serviceloc by names for < EDB2

zmousm committed:
fixup! Reinstate ServiceLoc.{latitude,longitude} as cached property
descriptors

* Fix 'set' object does not support indexing

zmousm committed:
Ensure Coordinates are linked to a single ServiceLoc object through a
signal

* Add an m2m_changed signal receiver to enforce this
* It works like edumanage.signals.sloc_coordinates_enforce_one(), but in
  the oppposite direction
* Use a separate receiver, as this will stick around even after the
  single Coordinates per ServiceLoc restriction gets lifted

zmousm committed:
Clean to-be-orphaned Coordinates objects

When a ServiceLoc is removed, this does not trigger cascaded removal of
the linked Coordinates object due to the M2M relationship. Add a signal
receiver to take care of that.

zmousm committed:
edumanage.management.commands.parse_institution_xml: Refactor parsing
contacts

* A contact is matched only by name and the M2M relationship to the
  respective institution
  * Other parsed parameters update the respective fields
* Check all child element tags are valid and all expected tags are
  present

zmousm committed:
edumanage.management.commands.parse_institution_xml: Provide help for
edb_version option

zmousm committed:
edumanage.management.commands.parse_institution_xml: Refactor and
simplify matching a ServiceLoc

* Progressively filter by coordinates, names, addresses if more
  objects are matched at each step, only for < EDB2
  * Stop once results are narrowed down to 0 or 1
  * This provides the most accurate matching possible for EDB1
    locations (where locationid is not available)
* Use parse_name (and parse_address) function to build terms of field
  lookups for iterative filter chaining, rather than the elaborated
  set comparison

zmousm committed:
edumanage.management.commands.parse_institution_xml: Split parsing
for names and addresses

* Split parsing to separate functions for parse_and_create_name(),
  parse_and_create_address(), so that parsing can be reused where
  necessary
* Let these functions raise errors with the respective messages rather
  than logging them

zmousm committed:
edumanage.management.commands.parse_institution_xml: fixes

* Fix typo in parsing address lang attribute
* Use more specific exception handling for URL parsing
* Fix typo for settings.SERVICELOC_DERIVE_WIRED_NO in parsing location
* Check that parameter exists before fetching value to
  instdetails_defaults
* While enumerating coordinates, fall back to an empty list if
  institution has none

zmousm committed:
utils.edb_versioning.DelegateVersion: Ensure comparisons are not
excluded from delegated methods under PY3

Fix TypeError: '>=' not supported between instances of
'EduroamDatabaseVersion' and 'Version'

zmousm committed:
fixup! edumanage.views.ourPoints: Adapt for ServiceLoc.tag

zmousm committed:
Add a context manager to disable signal receivers by dispatch_uid and
use it in parse_institution_xml

* Use constants for the two signal receiver identifiers of interest
* Add context manager edumanage.signals.disable_signals that
  temporarily stashes aside signal receivers by dispatch_uid
* Use disable_signals to disable the two ServiceLoc post_save signal
  receivers while parse_and_create_serviceloc() is called
* Manually trigger edumanage.ourPoints at the end, once per
  Institution rather than upon saving every ServiceLoc

zmousm commented:
I am now pushing the final round of fixes, after testing `parse_institution_xml`.

zmousm commented:
Thank you for your efforts, I understand testing and/or reviewing is not easy for this particularly large change-set.

I will squash all _fixup!_ commits and merge this shortly due to (external) time constraints.

vladimir-mencl-eresearch commented:
Hi @zmousm ,

Thanks for all the effort put into it - and sorry about the slow action on my side.
I've been trying to go through the second batch of commits, but was not progressing much - and there was a lot to follow, sometimes out of my depth with Django quirks, sometimes just plain lots of changes.  But overall, it all looked reasonable.

If you'd like me to at least help testing, please let me know when ready to go (and which branch).

Cheers,
Vlad

zmousm commented:
Hi @vladimir-mencl-eresearch,

throughout this effort I tried to only touch code that had to be updated for EDB v2, yet avoid working around code that needed refactoring. DjNRO is an 8 year old project, so the latter was all over the place :) I am sorry if I made it harder for you.

What you can do is deploy the code in master (considering the notes/caveats provided previously) and check if it works out for any niche use cases you may have in mind -- maybe you will find a bug I didn't catch!
